### DISCUSSION TITLE
How do I secure my Node.js app with JWT authentication?

### DISCUSSION THREAD

**Discussion by IT22271532** [community/community#142399] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

 I'm building a **Node.js** application and I want to implement authentication using JWT (JSON Web Tokens). I understand the basics of JWT but would like to know how to securely implement it in a Node.js environment.

How should I generate JWTs, protect routes, and manage tokens to ensure that my application is secure? Additionally, how do I handle password hashing when using **JWT-based authentication**?

**Comment by madhawaawishka** [comment]
To secure your Node.js app with JWT authentication, follow these steps:

1.**Install Required Packages**
First, install the necessary dependencies like **jsonwebtoken** for **JWT** and **bcryptjs** for password hashing.
`npm install jsonwebtoken bcryptjs express`

2.**Generate a JWT on User Login**
When a user logs in, create a token with jsonwebtoken. Use the user's unique information (such as their username or ID) to generate the token, and set an expiration time for security.

`const jwt = require("jsonwebtoken");`

`const loginUser = (req, res) => {
  const { username, password } = req.body;`

  // Verify user's credentials (password hashing can be done using bcrypt)
  // After verifying the user's details...

  // Create JWT token
  `const token = jwt.sign({ username }, process.env.JWT_SECRET, {
    expiresIn: "1h"`, // Expiration time of 1 hour
  });

  // Send token back to the client
 ` res.json({ token });`
`};`

3.**Protect Routes with JWT Middleware**
Use middleware to verify the token for protected routes. This middleware will check if the token is present and valid before allowing access.
`const verifyToken = (req, res, next) => {
  const token = req.headers["authorization"];`

  `if (!token) {
    return res.status(403).send("A token is required for authentication");
  }`

  `try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    req.user = decoded;
  } catch (err) {
    return res.status(401).send("Invalid Token");
  }`

 ` return next();
};
`

Apply this middleware to any route you want to protect:

`app.get("/dashboard", verifyToken, (req, res) => {
  res.send("This is a protected route");
});
`

4.**Hash Passwords with bcrypt**
Securely hash passwords before saving them to the database to ensure that even if the database is compromised, passwords remain protected.
`const bcrypt = require("bcryptjs");`

`const registerUser = async (req, res) => {
  const { username, password } = req.body;`

  // Hash the password before storing it
  `const hashedPassword = await bcrypt.hash(password, 10);`

  // Store the hashed password in the database
  // Continue with registration logic...
`};
`

5.**Store JWT Secret Safely**
Store your JWT secret key securely in environment variables using a .env file:
`JWT_SECRET=yourSuperSecretKey
`

By implementing JWT authentication with these steps, you can ensure that your Node.js application is secure, with safe password handling and protected routes for sensitive information.

**Comment by ebndev** [comment]
Hi @IT22271532 & @madhawaawishka, thanks for being a part of the GitHub Community!

While we appreciate you wanting to contribute to Discussions, please note: We made the decision to disable the ability to earn Achievements in our Community in order to discourage users from participating in coordinated or inauthentic activity like rapid questions and answers in order to earn badges. You can learn more about this decision in our announcement post here [Achievements will no longer be available in the Community](https://github.com/orgs/community/discussions/106536).

As a result, we'll be unmarking the answer and locking this, and other related posts. 

Any future violations may result in a temporary or indefinite block from the Community. 

Thanks for understanding and we hope to see you participate in other discussions in the future!
